File: projectvoortgang/backend/app/schemas/opname.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class ProjectopnameCreate(BaseModel):
    hoofdproject_key: int
    hoofdproject: str | None = None
    hoogst_geselecteerd_projectniveau: int | None = None
    start_boekdatum: date | None = None
    einde_boekdatum: date | None = None
    grondslag_calculatie_kosten: str | None = None
    grondslag_geboekte_kosten: str | None = None
    groepering_paragraafniveau: int | None = None

    # Raw DWH aggregates (filled during creation from DWH data)
    calculatie_kostprijs_inkoop: float = 0
    calculatie_kostprijs_arbeid_montage: float = 0
    calculatie_kostprijs_arbeid_projectgebonden: float = 0
    calculatie_verrekenprijs_inkoop: float = 0
    calculatie_verrekenprijs_arbeid_montage: float = 0
    calculatie_verrekenprijs_arbeid_projectgebonden: float = 0
    calculatie_montage_uren: float = 0
    calculatie_projectgebonden_uren: float = 0
    definitieve_kostprijs_inkoop: float = 0
    definitieve_kostprijs_arbeid_montage: float = 0
    definitieve_kostprijs_arbeid_projectgebonden: float = 0
    definitieve_verrekenprijs_inkoop: float = 0
    definitieve_verrekenprijs_arbeid_montage: float = 0
    definitieve_verrekenprijs_arbeid_projectgebonden: float = 0
    onverwerkte_verrekenprijs_inkoop: float = 0
    onverwerkte_verrekenprijs_arbeid_montage: float = 0
    onverwerkte_verrekenprijs_arbeid_projectgebonden: float = 0
    montage_uren_definitief: float = 0
    montage_uren_onverwerkt: float = 0
    projectgebonden_uren_definitief: float = 0
    projectgebonden_uren_onverwerkt: float = 0
    historische_verzoeken_inkoop: float = 0
    historische_verzoeken_montage: float = 0
    historische_verzoeken_projectgebonden: float = 0
    historische_verzoeken_montage_uren: float = 0
    historische_verzoeken_projectgebonden_uren: float = 0


class ProjectopnameUpdate(BaseModel):
    hoofdproject: str | None = None
    start_boekdatum: date | None = None
    einde_boekdatum: date | None = None
    grondslag_calculatie_kosten: str | None = None
    grondslag_geboekte_kosten: str | None = None
    groepering_paragraafniveau: int | None = None
    opmerking: str | None = None


class ProjectopnameResponse(BaseModel):
    projectopname_key: int
    klantnummer: int
    hoofdproject_key: int
    hoofdproject: str | None
    hoogst_geselecteerd_projectniveau: int | None
    start_boekdatum: date | None
    einde_boekdatum: date | None
    grondslag_calculatie_kosten: str | None
    grondslag_geboekte_kosten: str | None
    groepering_paragraafniveau: int | None

    # Resolved values
    calculatie_inkoop: float
    calculatie_montage: float
    calculatie_projectgebonden: float
    geboekt_inkoop: float
    geboekt_montage: float
    geboekt_projectgebonden: float
    geboekt_montage_uren: float
    geboekt_projectgebonden_uren: float

    # Calculated
    tmb_inkoop: float
    tmb_montage: float
    tmb_projectgebonden: float
    tmb_montage_uren: float
    tmb_projectgebonden_uren: float

    verschil_inkoop_huidige_stand: float
    verschil_montage_huidige_stand: float
    verschil_projectgebonden_huidige_stand: float

    gemiddeld_pg_inkoop: float
    gemiddeld_pg_montage: float
    gemiddeld_pg_projectgebonden: float
    gemiddeld_pg_totaal: float

    verschil_inkoop_einde_project: float
    verschil_montage_einde_project: float
    verschil_projectgebonden_einde_project: float

    ondergrens_inkoop: float
    ondergrens_montage: float
    ondergrens_projectgebonden: float
    bovengrens_inkoop: float
    bovengrens_montage: float
    bovengrens_projectgebonden: float

    # Historical requests
    historische_verzoeken_inkoop: float
    historische_verzoeken_montage: float
    historische_verzoeken_projectgebonden: float

    # Status
    autorisatie_status: str
    opgeslagen: bool
    opmerking: str | None

    # Audit
    aanmaakdatum: datetime
    aanmaker: str
    wijzigdatum: datetime
    wijziger: str

    model_config = {"from_attributes": True}

File: projectvoortgang/frontend/src/types/opname.ts
export interface Projectopname {
  projectopname_key: number
  klantnummer: number
  hoofdproject_key: number
  hoofdproject: string | null
  hoogst_geselecteerd_projectniveau: number | null
  start_boekdatum: string | null
  einde_boekdatum: string | null
  grondslag_calculatie_kosten: string | null
  grondslag_geboekte_kosten: string | null
  groepering_paragraafniveau: number | null

  // Resolved
  calculatie_inkoop: number
  calculatie_montage: number
  calculatie_projectgebonden: number
  geboekt_inkoop: number
  geboekt_montage: number
  geboekt_projectgebonden: number
  geboekt_montage_uren: number
  geboekt_projectgebonden_uren: number

  // TMB
  tmb_inkoop: number
  tmb_montage: number
  tmb_projectgebonden: number
  tmb_montage_uren: number
  tmb_projectgebonden_uren: number

  // Verschil
  verschil_inkoop_huidige_stand: number
  verschil_montage_huidige_stand: number
  verschil_projectgebonden_huidige_stand: number

  // Gemiddeld PG
  gemiddeld_pg_inkoop: number
  gemiddeld_pg_montage: number
  gemiddeld_pg_projectgebonden: number
  gemiddeld_pg_totaal: number

  // Verschil einde project
  verschil_inkoop_einde_project: number
  verschil_montage_einde_project: number
  verschil_projectgebonden_einde_project: number

  // Grenzen
  ondergrens_inkoop: number
  ondergrens_montage: number
  ondergrens_projectgebonden: number
  bovengrens_inkoop: number
  bovengrens_montage: number
  bovengrens_projectgebonden: number

  // Historische verzoeken
  historische_verzoeken_inkoop: number
  historische_verzoeken_montage: number
  historische_verzoeken_projectgebonden: number

  // Status
  autorisatie_status: string
  opgeslagen: boolean
  opmerking: string | null

  // Audit
  aanmaakdatum: string
  aanmaker: string
  wijzigdatum: string
  wijziger: string
}

export interface ProjectopnameCreate {
  hoofdproject_key: number
  hoofdproject?: string
  start_boekdatum?: string
  einde_boekdatum?: string
  grondslag_calculatie_kosten: string
  grondslag_geboekte_kosten: string
  groepering_paragraafniveau?: number
}

export interface Opnameregel {
  regel_key: number
  projectopname_key: number
  klantnummer: number
  project_key: number | null
  project: string | null
  projectniveau: number | null
  projectfase: string | null
  deelproject_jn: string
  bestekparagraaf_key: number | null
  bestekparagraaf: string | null
  bestekparagraafniveau: number | null

  calculatie_kosten_inkoop: number
  calculatie_kosten_arbeid_montage: number
  calculatie_kosten_arbeid_projectgebonden: number
  calculatie_montage_uren: number
  calculatie_projectgebonden_uren: number

  geboekte_kosten_inkoop: number
  geboekte_kosten_arbeid_montage: number
  geboekte_kosten_arbeid_projectgebonden: number
  montage_uren: number
  projectgebonden_uren: number

  laatste_pg_inkoop: number | null
  laatste_pg_montage: number | null
  laatste_pg_projectgebonden: number | null

  percentage_gereed_inkoop: number
  percentage_gereed_arbeid_montage: number
  percentage_gereed_arbeid_projectgebonden: number
}

export interface RegelUpdate {
  regel_key: number
  percentage_gereed_inkoop?: number
  percentage_gereed_arbeid_montage?: number
  percentage_gereed_arbeid_projectgebonden?: number
}
